lib/ide/src/controller/module.rs

```diff
@@ -28,32 +56,161 @@ impl Location {
     }
 }
 
+
+
+// =========================
+// === Module Controller ===
+// =========================
+
 shared! { Handle
     /// State data of the module controller.
     #[derive(Debug)]
     pub struct Controller {
         /// This module's location.
-        location : Location,
+        location: Location,
+        /// The current module ast, used by synchronizing both module representations.
+        ast: Ast,
+        /// The id map of current ast
+        // TODO: written for test purposes, should be removed once generating id_map from AST will
+        // be implemented.
+        id_map: IdMap,
+        /// The File Manager Client handle.
+        file_manager: fmc::Handle,
+        /// The Parser handle
+        parser: SharedParser,
+        logger: Logger,
     }
 
     impl {
-        /// Create new Module controller for given location.
-        pub fn new(location:Location) -> Self {
-            Controller {location:location}
-        }
-
         /// Obtain clone of location.
         pub fn location(&self) -> Location {
             self.location.clone()
         }
+
+        /// Updates AST after code change.
+        pub fn apply_code_change(&mut self,change:&TextChangedNotification) {
+            let mut code        = self.code();
+            let replaced_range  = change.replaced_chars.clone();
+            let inserted_string = change.inserted_string();
+            let replaced_size   = Size::new(replaced_range.end - replaced_range.start);
+            let replaced_span   = Span::new(Index::new(replaced_range.start),replaced_size);
+
+            code.replace_range(replaced_range,&inserted_string);
+            apply_code_change_to_id_map(&mut self.id_map,&replaced_span,&inserted_string);
+            self.ast = self.parser.parse(code, self.id_map.clone()).unwrap();
```

Comment:
We should not unwrap here. While parser "should never" fail, this can be triggered pretty easily. Never trust external components. 

We should complain as much as possible and leave previous AST intact, in hope that future edits will fix it. (unless we have idea for better behavior here)